### ISSUE TITLE
mount provider changes order of existing entries

### ISSUE THREAD

**Issue by scarpe01** [chef/chef#2908]
chef-solo --version
Chef: 11.12.4

This resulted in some seriuous headaches the past few days. The problem is pretty clear, if mount sees that a resource should be enabled, but does not match current state, it disables then enables the resource. This changes the order of entries in fstab which can have side effects of changing when in the boot process filesystems are loaded.

Here is how we produced the problem:
Our initial config has /var and /var/log setup with default options out of the installer.
Our requirement is to add a few options.

After we add our options with the mount resource, /var/log is defined before /var, causing it to not mount until netfs runs. Since rsyslogd starts up at S12 (netfs is at S26), so it begins logging on /var and then its logs become invisiable when the correct /var/log is mounted over them.,

In short, if entries exist before the provider is called, it should either attempt to preserve that order, especially for volumes important to the system.

**Comment by lamont-granquist** [comment]
- you need to replicate this on chef 12, there's been changes to the mount provider since 11.12.4
- you need to include your distro and version since there's different mount providers with different behaviors

**Comment by scarpe01** [comment]
Platform is OL6. sorry I buried that in the text.

I will give chef12 a try but, looking at the current code on github, I can see the issue is still there on the master branch anyway:
https://github.com/chef/chef/blob/master/lib/chef/provider/mount/mount.rb lines 153 to 160, there is clearly no logic to do anything but an append

Thopugh as you say, there may be different providers etc, so I will give it a try with the latest anyway, its easy enough to try.

**Comment by lamont-granquist** [comment]
OL6 == Oracle 6.x?

**Comment by scarpe01** [comment]
from chef-12.0.3-1.x86_64:

Oh yah, sorry we use that shorthand alot. Latest OL6 fully updated 6.6 latest. 

Before:
..
/dev/mapper/rootvg-varlv /var                    ext4    defaults        1 2
/dev/mapper/rootvg-loglv /var/log                ext4    defaults        1 2
..
proc                    /proc                   proc    defaults        0 0

After:
...
/dev/mapper/rootvg-loglv /var/log                ext4    defaults        1 2
...
/dev/mapper/rootvg-varlv /var auto defaults 0 2
...

Confirmed.

**Comment by bby-bishopclark** [comment]
Can confirm this is still occurring.

OS:  EL7.7  (RHEL/CentOS and friends) 
Chef Client : 14.13.11

Results:
- all changed lines are created at the tail of fstab
- old lines ARE NOT REMOVED

Thus before:

```
/dev/mapper/misc-logs /var/log/ ext4 defaults 0 2
/dev/mapper/data-vcpulp /var/cache/pulp/ ext4 defaults 0 2
/dev/mapper/data-vlmongodb /var/lib/mongodb/ ext4 defaults 0 2
/dev/mapper/data-vlqpidd /var/lib/qpidd/ ext4 defaults 0 2
```

and after (trailing slashes on mountpoints removed in added lines):

```
/dev/mapper/misc-logs /var/log/ ext4 defaults 0 2
/dev/mapper/data-vcpulp /var/cache/pulp/ ext4 defaults 0 2
/dev/mapper/data-vlmongodb /var/lib/mongodb/ ext4 defaults 0 2
/dev/mapper/data-vlqpidd /var/lib/qpidd/ ext4 defaults 0 2
/dev/mapper/misc-logs /var/log ext4 defaults 0 2
/dev/mapper/data-vcpulp /var/cache/pulp ext4 defaults 0 2
/dev/mapper/data-vlmongodb /var/lib/mongodb ext4 defaults 0 2
/dev/mapper/data-vlqpidd /var/lib/qpidd ext4 defaults 0 2
```

I'm available for testing changes, but I'm limited to chef-client 14.x

**Comment by antima-gupta** [comment]
@bby-bishopclark , @scarpe01  Issue is not replicating, [here](https://github.com/chef/chef/blob/5477a96d7725f557502939fc99531b0806978044/lib/chef/provider/mount/mount.rb#L54) we are checking if an entry exist in `fstab` .Could you please try on latest chef version and let me know the more information to replicate this, if you are still facing the same.

**Comment by bby-bishopclark** [comment]
>> I'm available for testing changes, but I'm limited to chef-client 14.x

> try on [the] latest chef version and let me know ~~the~~ more information to replicate this, if you are still facing the same [issue].

I cannot:  I'm limited to 14.x due to the licensing trapdoor in v15+ .  I can follow-up with the CinC project once they're minimally ready (for me, that's a YUM repo), though.

**Comment by dheerajd-msys** [comment]
Hi @bby-bishopclark @scarpe01 Please check ^ and let us know if you are still facing issues with latest chef infra.

**Comment by bby-bishopclark** [comment]
>> I'm available for testing changes, but I'm limited to chef-client 14.x

> Hi @bby-bishopclark @scarpe01 Please check ^ and let us know if you are still facing issues with latest chef infra.

So, 14.15.6 , then?

**Comment by dheerajd-msys** [comment]
@bby-bishopclark We've identified some related issue and looking forward to fix but it won't be available to Chef 14 as it's EOL now https://docs.chef.io/versions/#end-of-life-eol-products.
cc : @antima-gupta

**Comment by antima-gupta** [comment]
@bby-bishopclark On 16.x it's not making multiple entries but I can see multiple entries in  `/etc/fstab` on 15.x . Please let me know if you were facing the same issue.

## Recipe after making some change in mount resource:

```
mount "/mnt" do
  device "/home/abc/"
  fstype "ext4"
  options %w{bind rw}
  action %i{ mount enable }
end

mount "/mnt" do
  device "/home/bcd/"
  fstype "ext4"
  options %w{bind ro}
  action %i{ mount enable }
end
```

### Entry in `fstab` before: 

```
/home/abc/ /mnt ext4 bind,ro 0 2
/home/bcd/ /mnt ext4 bind,rw 0 2
```
### Entry in`fstab`  after running above recipe:

```
/home/abc/ /mnt ext4 bind,ro 0 2
/home/bcd/ /mnt ext4 bind,rw 0 2
/home/abc/ /mnt ext4 bind,rw 0 2
/home/antima/bcd/ /mnt ext4 bind,ro 0 2

```

**Comment by tas50** [comment]
This is resolved in 16.7
